File: 백준/Silver/2644. 촌수계산/촌수계산.py
import sys
input = sys.stdin.readline

n = int(input())
a, b = map(int, input().split())
m = int(input())
result = []

graph = [[] for _ in range(n + 1)]
visit = [False for _ in range(n + 1)]

for i in range(m):
    x, y = map(int, input().split())
    graph[x].append(y)
    graph[y].append(x)

def dfs(x, num):
    num += 1
    visit[x] = True

    if x == b:
        result.append(num)

    for i in graph[x]:
        if not visit[i]:
            dfs(i, num)

dfs(a, 0)

if len(result) == 0:
    print(-1)
else:
    print(result[0] - 1)

File: 백준/Silver/2644. 촌수계산/촌수계산.java
import java.util.*;
import java.io.*;

class Main {
    static int n;
    static int start, end;
    static List<Integer>[] arr;
    static boolean[] visited;
    static int result = -1;
    
    public static void dfs(int idx, int cnt) {
        visited[idx] = true;
        
        for(Integer i : arr[idx]) {
            if(!visited[i]){
                if(end == i) {
                    result = cnt + 1;
                    return;
                }
                dfs(i, cnt + 1);
            }
        }
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        n = Integer.parseInt(br.readLine());
        arr = new List[n + 1];
        visited = new boolean[n + 1];
        
        StringTokenizer st = new StringTokenizer(br.readLine());
        start = Integer.parseInt(st.nextToken());
        end = Integer.parseInt(st.nextToken());
        
        for(int i = 0; i <= n; i++) {
            arr[i] = new ArrayList<>();
        }
        
        int num = Integer.parseInt(br.readLine());
        for(int i = 0; i < num; i++) {
            st = new StringTokenizer(br.readLine());
            int parent = Integer.parseInt(st.nextToken());
            int child = Integer.parseInt(st.nextToken());
            arr[parent].add(child);
            arr[child].add(parent);
        }
        
        dfs(start, 0);
        
        System.out.println(result);
    }
}